components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/OAuthAppTenantResolverValve.java

```diff
@@ -0,0 +1,157 @@
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.identity.oauth.common.exception.InvalidOAuthClientException;
+import org.wso2.carbon.identity.oauth.dao.OAuthAppDO;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.client.authentication.OAuthClientAuthnException;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+
+import java.io.IOException;
+
+import javax.servlet.ServletException;
+
+import static org.wso2.carbon.identity.oauth.common.OAuthConstants.OAuth10AParams.OAUTH_CONSUMER_KEY;
+import static org.wso2.carbon.identity.oauth.common.OAuthConstants.OAuth20Params.CLIENT_ID;
+import static org.wso2.carbon.identity.oauth.common.OAuthConstants.TENANT_NAME_FROM_CONTEXT;
+
+/**
+ * This valve is used to resolve the tenant domain of the OAuth application.
+ *
+ * When tenant qualified urls are not enabled, we need to set the tenant domain of the oauth app to the thread
+ * local. This is because with the client id tenant uniqueness improvement, DAO layer requires the tenant domain
+ * and client id to retrieve an app and the tenant is not available in the request path. Note that when tenant
+ * qualified urls are disabled, client id is unique across the server.
+ */
+public class OAuthAppTenantResolverValve extends ValveBase {
+
+    private static final Log log = LogFactory.getLog(OAuthAppTenantResolverValve.class);
+    private static String oAuthServerBaseURL = null;
+    private static String oAuth2ServerBaseURL = null;
+
+    @Override
+    public void invoke(Request request, Response response) throws IOException, ServletException {
+
+        try {
+            if (isOAuthRequest(request)) {
+                String appTenant = "";
+                String clientId = request.getParameter(CLIENT_ID);
+
+                if (StringUtils.isEmpty(clientId) && isOAuth10ARequest(request)) {
+                    clientId = request.getParameter(OAUTH_CONSUMER_KEY);
+                }
+
+                // If empty, try to get the client id from the authorization header.
+                if (StringUtils.isEmpty(clientId)) {
+                    if (OAuth2Util.isBasicAuthorizationHeaderExists(request)) {
+                        try {
+                            String[] credentials = OAuth2Util.extractCredentialsFromAuthzHeader(request);
+                            if (credentials != null) {
+                                clientId = credentials[0];
+                            }
+                        } catch (OAuthClientAuthnException e) {
+                            if (log.isDebugEnabled()) {
+                                log.debug("Error while extracting credentials from authorization header.", e);
+                            }
+                        }
+                    }
+                }
+
+                if (StringUtils.isNotEmpty(clientId)) {
```

Comment:
Don't we need to break the flow if the clientId is still empty.

Reply:
Resolving as per the offline chat and https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/240#discussion_r1338416845